File: 2d_debug/spline_map/src/spline_chain.ts
import { AABB } from "./aabb.js";
import { Plane } from "./plane.js";
import Spline from "./spline.js";
import { add, mat3, scale, subtract, transform, vec2 } from "./vec2.js";

class SplineChain {
    splines: Spline[];
    private amount: number;

    ts = vec2(0, 0);

    constructor(splines: Spline[]) {
        this.splines = splines;
        this.amount = splines.length;
    }

    static from_points(points: vec2[], tau: number = 0.2): SplineChain {
        points.unshift(points[0]);
        points.push(points[points.length - 1]);

        const splines = [];
        for (let index = 1; index < points.length - 2; index++) {
            splines.push(
                Spline.with_control_points(
                    points[index],
                    points[index + 1],
                    points[index - 1],
                    points[index + 2],
                    tau
                )
            );
        }

        return new SplineChain(splines);
    }

    static from_points_with_outer_tangents(first_tangent: vec2, last_tangent: vec2, points: vec2[], tau: number = 0.2): SplineChain {
        points.unshift(subtract(points[0], first_tangent));
        points.push(add(points[points.length - 1], last_tangent));

        const splines = [];
        for (let index = 1; index < points.length - 2; index++) {
            splines.push(
                Spline.with_control_points(
                    points[index],
                    points[index + 1],
                    points[index - 1],
                    points[index + 2],
                    tau
                )
            );
        }

        return new SplineChain(splines);
    }

    static from_points_with_tangents(points: vec2[], tangents: vec2[]): SplineChain {
        if (points.length !== tangents.length) throw "Unequal amount of points and tangents";

        const splines = [];
        for (let index = 0; index < points.length - 1; index++) {
            splines.push(
                Spline.with_tangents(
                    points[index],
                    points[index + 1],
                    tangents[index],
                    tangents[index + 1]
                )
            );
        }

        return new SplineChain(splines);
    }

    static transform(spline_chain: SplineChain, matrix: mat3): SplineChain {
        return new SplineChain(spline_chain.splines.map(spline => Spline.transform(spline, matrix)));
    }

    get_extremes(): Float32Array {
        const results: number[] = [];
        
        for (let i = 0; i < this.splines.length; i++) {
            this.splines[i].get_extremes().forEach(value => results.push((value + i) / this.amount));
        }

        return Float32Array.from(results);
    }

    intersect_spline_plane(plane: Plane): boolean {
        const transformed_min = transform(plane.min, plane.inv_matrix);
        const transformed_max = transform(plane.max, plane.inv_matrix);

        let intersection: boolean = false;

        for (let i = 0; i < this.splines.length; i++) {
            const transformed_spline = Spline.transform(this.splines[i], plane.inv_matrix);
            intersection ||= transformed_spline.intersect_spline_aabb(new AABB(transformed_min, transformed_max));
            this.ts = scale(add(transformed_spline.ts, vec2(i, i)), 1.0 / this.amount);

            if (intersection) break;
        }

        return intersection;
    }

    intersect_spline_chain_plane_transformed(y: number): [boolean, number] {
        return this.splines[0].intersect_spline_plane_transformed(y);
    }

    draw_point_at(ctx: CanvasRenderingContext2D, t: number): void {
        const point = this.position_on_chain(t);

        ctx.beginPath();
        ctx.arc(
            Math.max(-ctx.canvas.width / 2, Math.min(ctx.canvas.width / 2, point.x)),
            Math.max(-ctx.canvas.height / 2, Math.min(ctx.canvas.height / 2, point.y)),
            0.02, 0.0, 2.0 * Math.PI);
        ctx.fill();
        ctx.stroke();
    }

    draw(ctx: CanvasRenderingContext2D, step: number = 0.01): void {
        for (const spline of this.splines) {
            spline.draw(ctx, step);
        }
    }

    position_on_chain(t: number): vec2 {
        t = Math.max(0.0, Math.min(t, 1.0 - 1e-4));
        const index: number = Math.floor(t * this.amount);
        const t_prime: number = (t - index / this.amount) * this.amount;

        return this.splines[index].position_on_spline(t_prime);
    }
}

export default SplineChain;

File: src/spline_chain.cpp
#include "spline_chain.hpp"
#include <cassert>

SplineChain::SplineChain() {
    this->length = 1;
    this->splines[0] = Spline();
}

SplineChain::SplineChain(std::vector<Spline> splines) {
    assert(splines.size() < MAX_SPLINE_CHAIN_LENGTH);

    this->length = splines.size();
    std::copy(splines.begin(), splines.end(), this->splines);
}

void SplineChain::init_vao() {
    for (int i = 0; i < this->length; i++) {
        this->splines[i].init_vao();
    }
}

void SplineChain::render() const {
    for (int i = 0; i < this->length; i++) {
        this->splines[i].render();
    }
}

void SplineChain::update_buffers() {
    for (int i = 0; i < this->length; i++) {
        this->splines[i].update_buffers();
    }
}

void SplineChain::clean() {
    for (int i = 0; i < this->length; i++) {
        this->splines[i].clean();
    }
}

SplineChain SplineChain::from_points(std::deque<glm::vec3> points, float tau) {
    return SplineChain::from_points_with_outer_tangents(
        glm::vec3(0.0f),
        glm::vec3(0.0f),
        points,
        tau
    );
}

SplineChain SplineChain::from_points_with_outer_tangents(glm::vec3 first_tangent, glm::vec3 last_tangent, std::deque<glm::vec3> points, float tau) {
    assert(points.size() > 1);
    std::vector<Spline> splines;

    points.push_front(points.front() - first_tangent);
    points.push_back(points.back() + last_tangent);

    for (int i = 1; i < points.size() - 2; i++) {
        splines.push_back(
            Spline::with_control_points(
                points[i],
                points[i + 1],
                points[i - 1],
                points[i + 2],
                tau
            )
        );
    }

    return SplineChain(splines);
}

SplineChain SplineChain::from_points_with_tangents(std::vector<glm::vec3> points, std::vector<glm::vec3> tangents) {
    assert(points.size() > 1);
    std::vector<Spline> splines;

    for (int i = 1; i < points.size(); i++) {
        splines.push_back(
            Spline::with_tangents(
                points[i - 1],
                points[i],
                tangents[i - 1],
                tangents[i]
            )
        );
    }

    return SplineChain(splines);
}

glm::vec3 SplineChain::position_on_chain(const float t) const {
    const float clamped_t = std::max(0.0f, std::min(t, 1.0f - 1e-4f));
    const unsigned int index = (unsigned int) floor(clamped_t * float(this->length));
    const float t_prime = (clamped_t - float(index) / float(this->length)) * float(this->length);

    return this->splines[index].position_on_spline(t_prime);
}

glm::vec3 SplineChain::position_on_transformed_chain(const float t) const {
    const float clamped_t = std::max(0.0f, std::min(t, 1.0f - 1e-4f));
    const unsigned int index = (unsigned int) floor(clamped_t * float(this->length));
    const float t_prime = (clamped_t - float(index) / float(this->length)) * float(this->length);

    return this->splines[index].transformedSpline->position_on_spline(t_prime);
}

std::vector<float> SplineChain::get_extremes() const {
    std::vector<float> result;

    for (int i = 0; i < this->length; i++) {
        std::vector<float> temp = this->splines[i].get_extremes();
        for (float t : temp) {
            result.push_back((t + float(i)) / this->length);
        }
    }

    return result;
}

std::optional<float> SplineChain::intersect_spline_plane(const glm::vec3 p) const {
    for (int i = 0; i < this->length; i++) {
        std::optional<float> result = this->splines[i].intersect_spline_plane(p);
        if (result.has_value()) {
            return (result.value() + i) / this->length;
        }
    }

    return std::nullopt;
}

void SplineChain::intersect_spline_plane(const BetterPlane& p, float ts[3 * MAX_SPLINE_CHAIN_LENGTH]) const {
    glm::vec3 _ts;
    for (int i = 0; i < this->length; i++) {
        this->splines[i].intersect_spline_plane(p, _ts);
        ts[i * 3 + 0] = (_ts.x + i) / this->length;
        ts[i * 3 + 1] = (_ts.y + i) / this->length;
        ts[i * 3 + 2] = (_ts.z + i) / this->length;
    }
}

void SplineChain::with_transform(const BetterPlane& plane) {
    for (Spline& spline : this->splines) {
        spline.with_transform(plane);
    }
}
